### PR TITLE
Fixes for Python 3

### PR DESCRIPTION
**Pull request by fsonntag**
Two small fixes for Python 3. I ran into those issues by playing around with the seq2seq library.

The problems can be easily reproduced by running [the test script](https://github.com/farizrahman4u/seq2seq/blob/master/tests/test.py) of seq2seq. Those error can also arise from other usages.

I hope it's okay to directly open a pull request without reporting an issue first.

### CODE REVIEW

File: keras/engine/training.py [hunk 1]
```diff
@@ -258,7 +258,10 @@ def collect_trainable_weights(layer):
         weights += layer.trainable_weights
     # dedupe weights
     weights = list(set(weights))
-    weights.sort(key=lambda x: x.name)
+    try:
+        weights.sort(key=lambda x: x.name)
+    except TypeError:
+        pass
```
Comment (fsonntag) [inline keras/engine/training.py]: So in Python 3 the sort method throws a TypeError, when the `name` is `None`. This doesn't occur in Python 2, so one has to catch this exception.

Comment (fchollet) [inline keras/engine/training.py]: The weights should be sorted. Simply catching the exception is not the right fix...

File: keras/engine/training.py [hunk 2]
```diff
@@ -258,7 +259,14 @@ def collect_trainable_weights(layer):
         weights += layer.trainable_weights
     # dedupe weights
     weights = list(set(weights))
-    weights.sort(key=lambda x: x.name)
+    # TF variables have auto-generated the name, while Theano has auto-generated the auto_name variable. name in Theano is None
+    if weights:
```
Comment (RamaneekGill) [inline keras/engine/training.py]: This still seems hackey. Maybe a better or more clear situation could be something like:

```python
from keras import backend as K
if K._BACKEND == 'tensorflow':
   logic
else:
   other_log
```

File: keras/engine/training.py (hunk 1 above)
Comment (RamaneekGill) [inline keras/engine/training.py]: @fsonntag @fchollet I added a comment on the PR that _may_ make it less hacky

File: keras/engine/training.py (hunk 2 above)
Comment (fsonntag) [inline keras/engine/training.py]: Thanks, updated that accordingly.

File: keras/engine/training.py [hunk 3]
```diff
@@ -258,7 +259,16 @@ def collect_trainable_weights(layer):
...
+        if K._BACKEND == 'tensorflow':
+            weights.sort(key=lambda x: x.name)
+        elif K._BACKEND == 'theano':
+            # auto_name looks e.g. like this 'auto_11', so we have to parse the int and sort by that
```
Comment (fchollet) [inline keras/engine/training.py]: Actually we don't; what matters is that the weights are in a deterministic order. In *what* order they are is not very important. We could simply sort by `auto_name`.

File: keras/engine/training.py [hunk 4]
```diff
@@ -258,7 +259,16 @@ def collect_trainable_weights(layer):
         weights += layer.trainable_weights
     # dedupe weights
     weights = list(set(weights))
-    weights.sort(key=lambda x: x.name)
+    # TF variables have auto-generated the name, while Theano has auto-generated the auto_name variable. name in Theano is None
+    if weights:
+        if K._BACKEND == 'tensorflow':
```
Comment (fchollet) [inline keras/engine/training.py]: The proper way to check is via `K.backend() == 'tensorflow'`.

File: keras/engine/training.py [hunk 5]
```diff
@@ -258,7 +259,14 @@ def collect_trainable_weights(layer):
...
     weights = list(set(weights))
-    weights.sort(key=lambda x: x.name)
+    # TF variables have auto-generated the name, while Theano has auto-generated the auto_name variable. name in Theano is None
+    if weights:
+        if K.backend() == 'tensorflow':
+            weights.sort(key=lambda x: x.name)
+        elif K.backend() == 'theano':
+            weights.sort(key=lambda x: x.auto_name)
+        else:
+            raise Exception('Unknown backend: ' + str(K._BACKEND))
```
Comment (fchollet) [inline keras/engine/training.py]: Since this is about dealing with a Theano-specific behavior, the syntax should be:

if Theano:
else:

Comment (fsonntag) [inline keras/engine/training.py]: K, I adjusted that accordingly.
Nevertheless I'm not sure, if that's the way to deal with it. If another backend would be used, you don't know of sorting by `name` would work, right?

**Pull request merged**